src/hotspot/share/gc/g1/g1BatchedGangTask.hpp

```diff
@@ -0,0 +1,141 @@
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_GC_G1_G1BATCHEDGANGTASK_HPP
+#define SHARE_GC_G1_G1BATCHEDGANGTASK_HPP
+
+#include "gc/g1/g1GCPhaseTimes.hpp"
+#include "gc/shared/workgroup.hpp"
+#include "memory/allocation.hpp"
+
+template <typename E, MEMFLAGS F>
+class GrowableArrayCHeap;
+
+// G1AbstractSubTask represents a task to be performed either within an
+// G1BatchedGangTask running on a single worker ("serially") or multiple workers
+// ("in parallel"). A G1AbstractSubTask is always associated with a phase tag
+// that is used to automatically store timing information.
+//
+// A "serial" task is some piece of work that either can not be parallelized
+// easily, or is typically so short that parallelization is not worth the effort.
+// Current examples would be summarizing per worker thread information gathered
+// during garbage collection (e.g. Merge PSS work).
+//
+// A "parallel" task could be some large amount of work that typically naturally
+// splits across the heap in some way. Current examples would be clearing the
+// card table.
+//
+// See G1BatchedGangTask for information on execution.
+class G1AbstractSubTask : public CHeapObj<mtGC> {
+  G1GCPhaseTimes::GCParPhases _tag;
+
+  NONCOPYABLE(G1AbstractSubTask);
+
+protected:
+  // Record work item for this tag in G1GCPhaseTimes.
+  void record_work_item(uint worker_id, uint index, size_t count);
+
+public:
+  G1AbstractSubTask(G1GCPhaseTimes::GCParPhases tag) : _tag(tag) { }
+  virtual ~G1AbstractSubTask() { }
+
+  // How many workers (threads) would this task be able to keep busy for at least
+  // as long as to amortize worker startup costs.
+  // Called by G1BatchedGangTask to determine total number of workers.
+  virtual double num_busy_workers() const { return 1.0; }
+
+  // Called by G1BatchedGangTask to provide information about the the maximum
+  // number of workers for all subtasks after it has been determined.
+  virtual void set_max_workers(uint max_workers) { }
+
+  // Perform the actual work. Gets the worker id it is run on passed in.
+  virtual void do_work(uint worker_id) = 0;
+
+  // Tag for this G1AbstractSubTask.
+  G1GCPhaseTimes::GCParPhases tag() const { return _tag; }
+  // Human readable name derived from the tag.
+  const char* name() const;
+};
+
+// G1BatchedGangTask runs a set of G1AbstractSubTask using a work gang.
+//
+// Subclasses of this class add their G1AbstractSubTasks into either the list
+// of "serial" or the list of "parallel" tasks.
+// During execution in the work gang, this class will make sure that the "serial"
+// tasks are executed by a single worker only exactly once, but different "serial"
+// tasks may be executed in parallel using different workers. "Parallel" tasks'
+// do_work() method may be called by different workers passing a different
+// worker_id at the same time, but at most once per given worker id.
```

Comment:
worker id -> worker_id (I hope to one day rename worker_id to work_id, work_unit, or something else, but if we are not referring to the variable, we should not describe it as a worker id, as it does not describe a worker but a unit of work.